src/app/content/search/hooks/index.ts

```diff
@@ -52,45 +51,38 @@ export const receiveSearchHook: ActionHookBody<typeof receiveSearchResults> = (s
     : getFirstResult(book, payload);
 
   if (
-    // selectedResult may equal to null if api did not return any results
-    !selectedResult
+    selectedResult
     // We are clearing selected result when requesting a new search so in the theory this should never happen
-    || isEqual(select.selectedResult(state), selectedResult)
-    // selectedResult bookId data is different than current book id
-    || book.id !== getIndexData(selectedResult.result.index).bookId
+    && (isEqual(select.selectedResult(state), selectedResult)
+      // selectedResult bookId data is different than current book id
+      || book.id !== getIndexData(selectedResult.result.index).bookId)
   ) {
     return;
   }
 
-  services.dispatch(selectSearchResult(selectedResult));
-
-  const targetPageId = selectedResult.result.source.pageId;
-  const targetPage = assertDefined(
-    findArchiveTreeNodeById(book.tree, targetPageId),
-    'search result pointed to page that wasn\'t in book'
-  );
-
-  // currentPage may by undefined when user started a search from the 404 page
-  const page = currentPage || targetPage;
-
-  const navigation = {
-    params: getBookPageUrlAndParams(book, targetPage).params,
-    route: content,
-    state : {
-      bookUid: book.id,
-      bookVersion: book.version,
-      pageUid: stripIdVersion(targetPage.id),
-    },
-  };
-
-  const action = stripIdVersion(page.id) === stripIdVersion(targetPage.id) ? replace : push;
-  const search = queryString.stringify({
-    query,
-    target: JSON.stringify({ type: 'search', index: selectedResult.highlight }),
-  });
-  const hash = selectedResult.result.source.elementId;
+  if (selectedResult) {
+    services.dispatch(selectSearchResult(selectedResult));
+  }
+
+  const targetPageId = selectedResult?.result.source.pageId || currentPage?.id;
 
-  services.dispatch(action(navigation, { hash, search }));
+  const action = (targetPageId && stripIdVersion(targetPageId)) === (currentPage && stripIdVersion(currentPage.id))
+    ? replace : push;
+
+  const options = selectedResult
+    ? {
+      hash: selectedResult.result.source.elementId,
+      search: queryString.stringify({
+        query,
+        target: JSON.stringify({ type: 'search', index: selectedResult.highlight }),
+      }),
+    }
+    : { search: queryString.stringify({ query }) };
+
+  const navigationMatch = targetPageId ? createNavigationMatch(targetPageId, book) : selectNavigation.match(state);
+  if (navigationMatch) {
```

Comment:
the only reason this would be nullish would be if your `createNavigationMatch` helper decided the args were invalid right? in that case it'd be an error 

maybe that function should throw instead of returning undefined so you don't have to do null checks

Reply:
`selectNavigation.match(state)` returns `AnyMatch | undefined`

Reply:
hmmm... i think ideally `selectNavigation.match` would not be able to return undefined because it should always at least be the 404 route... but that isn't something we should look into on this pr.

for now lets make `createNavigationMatch` always return a match but leave the conditional afterwards

Reply:
This is how it is handled right now - it's ready for your review.